spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRequiredDistributionAndOrdering.java

```diff
@@ -0,0 +1,157 @@
+import org.junit.After;
+import org.junit.Test;
+
+public class TestRequiredDistributionAndOrdering extends SparkCatalogTestBase {
+
+  public TestRequiredDistributionAndOrdering(String catalogName, String implementation, Map<String, String> config) {
+    super(catalogName, implementation, config);
+  }
+
+  @After
+  public void dropTestTable() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testSortIdentityTransforms() throws NoSuchTableException {
+    sql("CREATE TABLE %s (c1 INT, c2 STRING, c3 STRING) " +
+        "USING iceberg " +
+        "PARTITIONED BY (c3)", tableName);
+
+    List<ThreeColumnRecord> data = ImmutableList.of(
+        new ThreeColumnRecord(1, null, "A"),
+        new ThreeColumnRecord(2, "BBBBBBBBBB", "B"),
+        new ThreeColumnRecord(3, "BBBBBBBBBB", "A"),
+        new ThreeColumnRecord(4, "BBBBBBBBBB", "B"),
+        new ThreeColumnRecord(5, "BBBBBBBBBB", "A"),
+        new ThreeColumnRecord(6, "BBBBBBBBBB", "B"),
+        new ThreeColumnRecord(7, "BBBBBBBBBB", "A")
+    );
+    Dataset<Row> ds = spark.createDataFrame(data, ThreeColumnRecord.class);
+    Dataset<Row> inputDF = ds.coalesce(1).sortWithinPartitions("c1");
+
+    // should succeed by default
+    inputDF.writeTo(tableName).append();
+
+    assertEquals("Row count must match",
+        ImmutableList.of(row(7L)),
+        sql("SELECT count(*) FROM %s", tableName));
+
+    Table table = validationCatalog.loadTable(tableIdent);
+
+    // should automatically prepend partition columns to the ordering
+    table.updateProperties()
+        .set(TableProperties.WRITE_DISTRIBUTION_MODE, TableProperties.WRITE_DISTRIBUTION_MODE_RANGE)
+        .commit();
+    table.replaceSortOrder()
+        .asc("c1")
+        .asc("c2")
+        .commit();
+    inputDF.writeTo(tableName).append();
+
+    assertEquals("Row count must match",
+        ImmutableList.of(row(14L)),
+        sql("SELECT count(*) FROM %s", tableName));
+
+    table.refresh();
+
+    // should succeed with a correct sort order
+    table.replaceSortOrder()
+        .asc("c3")
+        .asc("c1")
+        .asc("c2")
+        .commit();
+    inputDF.writeTo(tableName).append();
+
+    assertEquals("Row count must match",
+        ImmutableList.of(row(21L)),
+        sql("SELECT count(*) FROM %s", tableName));
+
+    // should fail if ordering is disabled
+    AssertHelpers.assertThrows("Should reject writes without ordering",
+        SparkException.class, "Writing job aborted",
+        () -> {
+          try {
+            inputDF.writeTo(tableName)
+                .option(SparkWriteOptions.DISTRIBUTION_MODE, "none")
+                .option(SparkWriteOptions.IGNORE_SORT_ORDER, "true")
+                .append();
+          } catch (NoSuchTableException e) {
+            throw new RuntimeException(e);
+          }
+        });
+  }
+
+  @Test
+  public void testNoSortOnBucketTransforms() throws NoSuchTableException {
```

Comment:
Can you add "WithoutExtensions" to the method name to make it clear why this fails?

Reply:
Added.